core/src/subgraph/loader.rs

```diff
@@ -0,0 +1,245 @@
+use std::collections::HashMap;
+use std::iter::FromIterator;
+
+use graph::data::subgraph::schema::*;
+use graph::data::subgraph::UnresolvedDataSource;
+use graph::prelude::{DataSourceLoader as DataSourceLoaderTrait, GraphQlRunner, *};
+use graph_graphql::graphql_parser::{parse_query, query as q};
+
+pub struct DataSourceLoader<L, Q, S> {
+    store: Arc<S>,
+    link_resolver: Arc<L>,
+    graphql_runner: Arc<Q>,
+}
+
+impl<L, Q, S> DataSourceLoader<L, Q, S>
+where
+    L: LinkResolver,
+    S: Store + SubgraphDeploymentStore,
+    Q: GraphQlRunner,
+{
+    pub fn new(store: Arc<S>, link_resolver: Arc<L>, graphql_runner: Arc<Q>) -> Self {
+        Self {
+            store,
+            link_resolver,
+            graphql_runner,
+        }
+    }
+
+    fn dynamic_data_sources_query(
+        &self,
+        deployment: &SubgraphDeploymentId,
+        skip: i32,
+    ) -> Result<Query, Error> {
+        // Obtain the "subgraphs" schema
+        let schema = self.store.subgraph_schema(&SUBGRAPHS_ID)?;
+
+        // Construct a query for the subgraph deployment and all its
+        // dynamic data sources
+        Ok(Query {
+            schema,
+            document: parse_query(
+                r#"
+                query deployment($id: ID!, $skip: Int!) {
```

Comment:
I really like the idea of using GraphQL here, makes the code much clearer than if this was implemented using store primitives. I am a little concerned about performance, since it's fairly deeply nested: do we have any data on how long this query will take with production-like amounts of data?

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a4802e01eeb5ea5ee699ac90c49d1cfa

Reply:
It looks like we should be fine in terms of number of SQL queries, the first two levels are by id, I think the worst path is `dynamicDataSources*templates*eventHandlers` which doesn't look like it will misbehave too much, probably not the dominating factor in starting a subgraph compared to resolving from IPFS.

Reply:
I ran a few manual SQL queries for parts of this query and they were all fast. This is no guarantee but I think I want to try this — on staging first, of course.